src/coreclr/jit/compiler.cpp

```diff
@@ -5518,12 +5521,39 @@ int Compiler::compCompile(CORINFO_MODULE_HANDLE classPtr,
 
     // Match OS for compMatchedVM
     CORINFO_EE_INFO* eeInfo = eeGetEEInfo();
-#ifdef TARGET_UNIX
-    info.compMatchedVM = info.compMatchedVM && (eeInfo->osType == CORINFO_UNIX);
-#else
-    info.compMatchedVM        = info.compMatchedVM && (eeInfo->osType == CORINFO_WINNT);
+
+#ifdef TARGET_OS_RUNTIMEDETERMINED
+    if (!TargetOS::OSSettingConfigured)
+    {
+        CORINFO_EE_INFO* eeInfo       = eeGetEEInfo();
+        TargetOS::IsMacOS             = eeInfo->osType == CORINFO_MACOS;
+        TargetOS::IsUnix              = (eeInfo->osType == CORINFO_UNIX) || (eeInfo->osType == CORINFO_MACOS);
+        TargetOS::IsWindows           = eeInfo->osType == CORINFO_WINNT;
+        TargetOS::OSSettingConfigured = true;
```

Comment:
Is it safe for these to be set as statics? E.g., multi-thread safe?

Do we want to support changing the target on multiple subsequent JIT invocations? E.g., would we want to support merging SuperPMI method contexts for different target OS settings and then replaying them all one after another with the same loaded JIT? Having these be `static` seems dangerous.

I wonder if they should be on the JIT "host" api ICorJitHost.